wom/src/main/scala/wdl4s/wom/graph/Graph.scala

```diff
@@ -23,14 +23,28 @@ object Graph {
     * Assuming it validates, construct the Graph case class.
     */
   def validateAndConstruct(nodes: Set[GraphNode]): ErrorOr[Graph] = {
-    def goodLink(port: InputPort): ErrorOr[Unit] = {
+
+    def boolToErrorOr(bool: Boolean, msg: => String): ErrorOr[Unit] = if (bool) {
+      ().validNel
+    } else {
+      msg.invalidNel
+    }
+
+    def upstreamNodeInGraph(port: InputPort): ErrorOr[Unit] = {
       val upstreamOutputPort = port.upstream
+      boolToErrorOr(nodes.exists(_ eq upstreamOutputPort.graphNode), s"The input link ${port.name} is linked to a node outside the graph set (${upstreamOutputPort.name})")
+    }
+
+    def portProperlyEmbedded(port: GraphNodePort, portFinder: GraphNode => Set[_ <: GraphNodePort]): ErrorOr[Unit] = {
+      boolToErrorOr(portFinder(port.graphNode).exists(_ eq port), s"The port ${port.name} thinks it belongs to a Node (${port.graphNode}), but that Node doesn't think it owns it.")
+    }
+
+    def goodLink(port: InputPort): ErrorOr[Unit] = {
+      val upstreamNodeValidation = upstreamNodeInGraph(port)
+      val inputPortEmbeddedValidation = portProperlyEmbedded(port, g => g.inputPorts)
+      val upstreamPortEmbeddedValidation = portProperlyEmbedded(port.upstream, g => g.outputPorts)
 
-      if (nodes.contains(upstreamOutputPort.graphNode)) {
-        ().validNel
-      } else {
-        s"The input link ${port.name} is linked to a node outside the graph set (${upstreamOutputPort.name})".invalidNel
-      }
+      (upstreamNodeValidation |@| inputPortEmbeddedValidation |@| upstreamPortEmbeddedValidation) map { (_, _, _) => () }
```

Comment:
It didn't work here, I guess because the `|@|` creates a `CartesianBuilder` rather than an `ErrorOr`?